src/bootstrap.jl

```diff
@@ -58,42 +64,92 @@ function update!(stat::BootstrapPoisson, args::Any...)
             update!(replicate, args...)
         end
     end
+    stat.cache_is_dirty = true
+
+    return
+end
 
+function update!(stat::FrozenBootstrap, args::Any...)
+    error("Cannot update a FrozenBootstrap object")
     return
 end
 
 # TODO: Make this work with non-univariate statistics by taking marginal
 #       quantiles
 function state(stat::Bootstrap)
-    R = length(stat.replicates)
+    return stat.replicates
+end
 
-    for (i, replicate) in enumerate(stat.replicates)
-        stat.replicate_states[i] = state(replicate)
+function state(stat::FrozenBootstrap)
+    return stat.cached_state
+end
+
+# update cached_state' states if necessary and return their values
+function cached_state(stat::Bootstrap)
+    if stat.cache_is_dirty
+        for (i, replicate) in enumerate(stat.replicates)
+            stat.cached_state[i] = state(replicate)
+        end
+        stat.cache_is_dirty = false
     end
+    return stat.cached_state
+end
+
+function cached_state(stat::FrozenBootstrap)
+    return stat.cached_state
+end
+
+function ci(stat::Bootstrap, α=0.05, method="quantile")
+    states = cached_state(stat)
 
     # If any NaN, return NaN, NaN
-    if any(isnan, stat.replicate_states)
+    if any(isnan, states)
         return (NaN, NaN)
     else
-        return (
-            quantile(stat.replicate_states, stat.α / 2),
-            quantile(stat.replicate_states, 1 - stat.α / 2),
-        )
+        if method == "quantile"
+            return (
+                quantile(states, α / 2),
+                quantile(states, 1 - α / 2),
+            )
+        elseif method == "normal"
+            norm_approx = Distributions.Normal(
+                mean(states),
+                std(states)
+            )
+            return (
+                quantile(norm_approx, α / 2),
+                quantile(norm_approx, 1 - α / 2)
+            )
+        else
+            error("Unrecognized confidence interval type: ", ci)
+        end
     end
 end
 
 nobs(stat::Bootstrap) = stat.n
 
+# Assumes a and b are independent.
+function -(a::Bootstrap, b::Bootstrap)
```

Comment:
This function won't be callable in most modules because it needs to be `Base.(:-)` to safely extend the core Julia definition of subtraction.